### ISSUE TITLE
[Language RFC] One floating-point reduction semantics: uniform min/max lowering plus scoped reassociation

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1082]
Baselines: Align `8c8bfbc7a3169e84ecc8415f5149ab8c61afe863` (alignc 0.7.5, LLVM 22.1.8), align-llm
`b72b0b48`, Apple M1, `--profile release`, default `--target-cpu baseline`. All IR/asm/timings below were
produced with that exact compiler on that host.

## Problem

Align has three different lowerings for "the maximum of these floats" and one lowering for "the sum of
these floats", and the differences are not explained by semantics. Two distinct issues hide in the same
place, and they need one design, not two patches:

**Part 1 — a lowering inconsistency that costs 16× today and requires no new semantics.** Align's scalar
`a.max(b)` already lowers to `llvm.maximum`, which LLVM 22 recognizes as a reduction and widens to
`fmax.4s` with no fast-math whatsoever. The **pipeline** `slice<f32>.max()` terminal and the hand-written
`if xs[i] > best` loop both lower to `fcmp ogt` + `select`, which LLVM cannot form a reduction from
because a `select` has no NaN semantics attached. Same mathematical operation, same NaN behaviour
available, 16× different throughput.

**Part 2 — a genuinely missing surface.** An additive or multiplicative float reduction (`sum`, `dot`,
`a*b+c` contraction) cannot be widened at all without reassociation, which Align correctly never does
silently. Align is at exact parity with clang under strict IEEE here, so this is not a codegen defect —
the opt-in simply does not exist. There is no `reassoc`, `contract`, `nnan` or `ninf` anywhere in the
compiler.

### Part 1: three lowerings of one operation

`crates/align_codegen_llvm/src/lib.rs:11461-11476` — the scalar math function, with its rationale:

```rust
                    align_sema::MathFn::Min | align_sema::MathFn::Max => {
                        let is_max = matches!(fn_, align_sema::MathFn::Max);
                        let name = if is_float {
                            // `minimum`/`maximum` (IEEE 754-2019), not `minnum`/`maxnum`: they
                            // propagate NaN and order ±0 deterministically — consistent across
                            // targets (Align values predictable, identical-across-builds results).
                            if is_max { "llvm.maximum" } else { "llvm.minimum" }
```

`crates/align_codegen_llvm/src/lib.rs:21320-21330` — the explicit vector lane reduction agrees:

```rust
        let name = if matches!(elem, Ty::Float(_)) {
            if max { "llvm.vector.reduce.fmaximum" } else { "llvm.vector.reduce.fminimum" }
```

But the pipeline reducer disagrees. `crates/align_mir/src/lib.rs:14403-14412` (and the same shape again at
`:14679-14690`) lowers `min`/`max` as a comparison plus a select:

```rust
            let op = if *is_max { BinOp::Gt } else { BinOp::Lt };
            let cmp = b.fresh_value(Ty::Bool);
            b.push(Stmt::Let(
                cmp,
                Rvalue::Bin(op, cur.clone(), Operand::Value(a)),
            ));
```

Three kernels, same input, emitted with the pinned `alignc` — identical loops differing only in the
maximum's spelling:

```align
pub fn max_cmp(borrow xs: slice<f32>) -> f32 {      // hand loop:   if xs[i] > best { best = xs[i] }
pub fn max_mathfn(borrow xs: slice<f32>) -> f32 {   // scalar term: best = best.max(xs[i])
pub fn max_pipe(borrow xs: slice<f32>) -> f32 = xs.max()   // pipeline terminal
```

`--stage optimized`. `max_cmp` and `max_pipe`:

```llvm
  %fgt = fcmp ogt float %slcload, %_1.035
  %_1.1 = select i1 %fgt, float %slcload, float %_1.035
```

— scalar, no `vector.body`. `max_mathfn` gets a two-way-unrolled vector reduction:

```llvm
  %6 = tail call <4 x float> @llvm.maximum.v4f32(<4 x float> %vec.phi, <4 x float> %wide.load)
  %7 = tail call <4 x float> @llvm.maximum.v4f32(<4 x float> %vec.phi27, <4 x float> %wide.load28)
```

and, at `emit-obj --target-cpu baseline`, `fmax.4s` ×3 plus a closing `fmaxv.4s s0, v0`, while
`max_cmp` and `max_pipe` get `fcmp s1, s0` + `fcsel s0, s1, s0, gt`.

Measured (152,064 `f32`, 300 calls, best of 3, same host, same objects):

| form | µs/call | GB/s |
|---|---|---|
| `if xs[i] > best { best = xs[i] }` | 191.0 | 3.18 |
| `xs.max()` pipeline terminal | 190.7 | 3.19 |
| `best = best.max(xs[i])` | **12.0** | **50.7** |

For reference, a clang `-O2` strict-IEEE C twin of the same reduction measures ~190.5 µs — i.e. Align's
pipeline terminal is at clang-strict parity, and `llvm.maximum` already beats it 16× *without relaxing
anything*. NaN propagation and ±0 ordering are unchanged, because `fmax`/`llvm.maximum` is precisely the
deterministic IEEE 754-2019 operation the scalar path already chose.

This also contradicts the language's own documentation: `examples/vec_simd.align:5` claims
*"automatic vectorization — the pipeline (`map`/`where`/`sum`) already vectorizes the same way"*,
which is true for `sum` in form but false in effect (below) and false for `max`/`min` in both.

### Part 2: additive reductions are at strict-IEEE parity, and the opt-in is missing

`git grep -n -E 'reassoc|nnan|ninf|fmuladd|set_fast_math|FastMathFlags|fpmath' -- \
'crates/align_codegen_llvm/src'` returns zero relevant hits. Float ops are plain
`build_float_add/mul` (`crates/align_codegen_llvm/src/lib.rs:21404-21407`, `:21345-21349`);
`llvm.fma` appears only for an explicit user `fma(a,b,c)` (`:11486`); `horizontal_sum`
(`:21298-21307`) calls `llvm.vector.reduce.fadd` **with a start value and no `reassoc`** — the
strictly ordered variant.

A hand `t = t + xs[i]` loop and the `xs.sum()` terminal emit **identical** IR — a widened load
feeding a serial ordered reduce, i.e. no arithmetic throughput at all:

```llvm
  %wide.load = load <4 x float>, ptr %2, align 1
  %3 = tail call float @llvm.vector.reduce.fadd.v4f32(float %vec.phi, <4 x float> %wide.load)
```

`out[i] = out[i] + a[i]*k` keeps a separate `fmul` + `fadd` and never contracts to `fmla`; the
align-llm release image contains **zero** `fmla` instructions anywhere.

| kernel (4096 f32) | Align | clang strict IEEE | clang `reassociate(on) contract(fast)` |
|---|---|---|---|
| f32 sum | 0.2796 s | 0.2758 s | **0.0160 s (17.5×)** |
| f32 dot | 0.2827 s | 0.2783 s | **0.0211 s (13.4×)** |

Across all `explain-opt` modules of align-llm the reassociation barrier is reported exactly **once**,
at `moe_layer_forward.align:2561:23` — not because it is rare, but because almost every other float
loop is blocked earlier by the view-header reload (#1079) and the surviving bounds checks
(#1081). The ordered `fadd` is visible in `model_forward$digest_region` `bb8`: scalar `fadd float`,
no vector body, 152,064 iterations per generated token.

### Why the relaxation must be opt-in and scoped, as a general language matter

Bit-for-bit reproducibility across builds, targets and thread counts is a first-class requirement for
some programs (a numeric oracle, a differential test against a reference implementation, a checksum
over a computed tensor, a financial accumulation) and irrelevant to others that only want throughput.
Neither can be inferred from the code, so the answer is not "relax by default" and not "never relax":
it is **ordered by default, relaxed only inside an explicit scope** — already settled at
`docs/open-questions.md:4956` (*"ordered floating-point reduction must not be silently
reassociated"*).

Evidence that the ordered requirement is real: `align-llm/src/layer_forward.align:2166` and
`moe_layer_forward.align:2561` accumulate in element order because their gates compare bit for bit
against a reference runtime:

```align
      // Sequential f32 accumulation in element order, which is section 2.3's finding: the
      // instrument accumulates its `sum` this way, and matching it made three of five checked nodes
      // bit-identical instead of merely close.
      accumulator = accumulator + region.f32_le(element * 4)
```

## Root cause

1. The pipeline `min`/`max` reducer was written as a generic `Bin(Gt/Lt)` + `Select` before
   `MathFn::Min`/`Max` chose `llvm.minimum`/`llvm.maximum`, and was never brought into line. Three code
   paths, three lowerings, one operation.
2. There is no surface anywhere in the language for "this reduction may be reassociated", so the 13–18×
   behind `reassoc`/`contract` is unreachable by any spelling.

## Proposed solution

One coherent floating-point reduction semantics, designed as a single surface rather than a set of named
terminals.

**(1) Unify min/max lowering (no new semantics, no new surface).** Route the pipeline `min`/`max` reducer
at `crates/align_mir/src/lib.rs:14403` and `:14679` through the same `MathFn::Min`/`MathFn::Max` path the
scalar method and the vector lane reduction already use, so all three spellings emit
`llvm.minimum`/`llvm.maximum`. NaN propagation, ±0 ordering and cross-target determinism are
*unchanged* — that is exactly what `llvm.maximum` guarantees and what the existing comment at
`lib.rs:11463-11466` already committed to. This is a "one way to do things" repair; it is not an
optimization exception.

**(2) One floating-point semantics model: ordered by default, with named relaxation levels in an
explicit lexical scope.** One mechanism applied uniformly to all float arithmetic, not
`sum_reassoc()`/`dot_reassoc()`/`max_fast()` one-offs: a family of named terminals grows without bound
(next `mean`, `variance`, `l2_norm`, `cumsum`) and leaves hand-written loops unanswered — the same
objection already raised against ad-hoc primitives on #1064. The model should state, once, for the
whole language:

- **default**: strictly ordered, IEEE 754, no reassociation, no contraction, NaN-propagating
  `minimum`/`maximum`. This is today's behaviour and does not change.
- **an explicit lexical scope** (e.g. a `@float(...)`-style attribute on a `fn` or block) naming which
  guarantees are relinquished, with each level independently selectable and separately documented:
  - `reassoc` — additive/multiplicative reductions may be reordered; selects the unordered
    `llvm.vector.reduce.fadd`/`fmul`;
  - `contract` — `a*b + c` may become one `fma`.

  Both are *result-defined*: every input still produces a defined float, only the association or the
  rounding of the intermediate changes. **`nnan` and `ninf` are deliberately excluded**, permanently
  and not as a later level: they make the result poison for a NaN or infinite input, which would put
  a hidden undefined-behaviour class inside a language whose settled rules are *"floats follow IEEE
  754 and never abort"* and *"nothing hidden"*. An input-assertion flag is also not needed for
  min/max once Part 1 lands, which is the only place the audit wanted one.
- **scoping rules**: lexical, non-inheriting across function boundaries (a callee's semantics never change
  because of its caller), applying uniformly to hand-written loops, pipeline terminals and explicit `vecN`
  arithmetic alike, so there is one rule and nothing to memorize per terminal.
- **outside any scope**: bit-for-bit unchanged, and that is testable
  (`crates/align_driver/tests/vectorize_shapes.rs:258` is already the negative control).

Level names and spelling are a design question; the requirement is that there is one surface, that it is
visible at the use site, and that each relinquished guarantee is named rather than bundled into a single
"fast" switch.

## Acceptance criteria

- `xs.max()` / `xs.min()` over `slice<f32>` emit `llvm.maximum`/`llvm.minimum` and a `<4 x float>`
  `vector.body` at `--profile release --target-cpu baseline`; `fmax.4s` + `fmaxv.4s` appear in the object.
  Measured: ≤ 20 µs/call on 152,064 `f32` (today 190.7 µs).
- `if xs[i] > best { best = xs[i] }` is documented as *not* the vectorizable spelling, or is canonicalized
  in MIR to the same `MathFn::Max` form; either answer is fine, but the choice is stated, and
  `examples/vec_simd.align` is corrected to match reality.
- A NaN/±0 conformance test shows the pipeline terminal's results are unchanged by (1) — including a NaN
  input, a `-0.0`/`+0.0` pair, and an all-NaN slice.
- Inside the Part-2 scope: `sum`/`dot` over 4096 `f32` emit an unordered `llvm.vector.reduce.fadd` and an
  `fmla`-carrying body, and land within ~1.5× of clang `reassociate(on) contract(fast)`.
- Outside the scope: `crates/align_driver/tests/vectorize_shapes.rs:258`
  (`k6_float_sum_does_not_vectorize_without_fast_math`) keeps passing unchanged for plain `+`, and
  `align-llm`'s bit-identity gates against llama.cpp still pass with the un-annotated accumulators at
  `src/layer_forward.align:2166` / `src/moe_layer_forward.align:2561`.
- Real-client target: `runtime_generation$greedy` over its 151,936 logits under **60 µs/token**
  (today 242 µs; the sampler is 267 µs over 152,064).

## Relation to existing issues/plans

- **Part 1 is new and was not in any prior report or issue.** It is a defect, not a feature request, and
  it is independently landable.
- Part 2 is a known, never-built backlog item: `docs/open-questions.md:5960` —
  *"Fast-math flags on float ops (opt-in): unlock float reassociation / autovectorization"*. The
  anti-implicit rule is settled at `docs/open-questions.md:4956` and
  `docs/impl/62-decode-optimization-plan.md:152-155` (*"An exact IEEE reduction may remain scalar."*), so
  the *direction* is pre-approved and only the *surface* is missing.
- `docs/impl/62-decode-optimization-plan.md:47` observes that *"Typed `slice<f32>.max()` in the
  baseline probe remains scalar too: type admission alone does not establish a legal profitable vector
  reduction."* Part 1 identifies *why* it remained scalar and shows it was legal all along.
- Directly relevant to **#1064** (revised to Option A, safe typed views), whose acceptance criterion
  *"normal loops over the resulting `slice<f32>` auto-vectorize cleanly"* cites `fmaxnm`/`vmaxps`.
  Part 1 is what makes that criterion achievable for `max`/`min`, with the caveat that Align's chosen
  instruction is `fmax`/`llvm.maximum` (NaN-propagating), not `fmaxnm`. The comment on #1064 records
  this.
- Also requires the loop-facts contract for the *access* half: *"[Codegen] Materialize a borrowed
  slice/array view header once and state its alias facts"* (#1079), *"[Codegen] Attach `!range` to
  slice/array length loads so a length is known non-negative"* (#1080) and *"[MIR] Fuse the bounds
  check into one unsigned compare and eliminate monotone-induction in-loop checks"* (#1081).
  Reassociation alone does not help a loop that is blocked earlier.
- Not a duplicate of #1063 (`exp` intrinsic) or #1047.

## Impact framing

Honest scope: in align-llm's current configuration the numeric work is on the GPU and Align-compiled
code is ~0.1 % of sampled CPU during Metal decode, so this is not a token-throughput claim. What it
does reach is the Align-owned CPU kernels (greedy logits selection 242 µs/token over 151,936 logits,
sampler 267 µs/token over 152,064, this host), the KV-plane verification and digest paths
(`model_forward$digest_region` runs an ordered 152,064-element `fadd` per generated token), and the
future in-language CPU inference path. Part 1 stands on its own as a design-consistency item: three
lowerings of one operation is a "one way to do things" violation regardless of performance.

**Comment by sanohiro** [comment]
Part 2 design is complete in #1144 (merge `69543a235e4764194bab04cfaf5ca3ca6702ca5c`). The issue remains open for implementation.

Acceptance criteria closed by this design merge:

- strict ordered/uncontracted floating-point behavior remains the default;
- `float(reassoc)`, `float(contract)`, and their explicit combination are the only relaxation surface;
- every callable root starts strict, while generic and interface-carried bodies recheck the exact source scope;
- checked HIR authenticates modes and the transparent wrapper preserves ownership, Drop, region/escape, effects, and every control-flow path;
- all five existing floating sum/dot variants (`ArraySum`, `ArrayDot`, `VecSum`, `VecSumWhere`, and `VecDot`) have exact mode semantics and owners;
- contraction never uses LLVM's consumer-controlled raw `contract` flag: Align selects only a direct two-sided authenticated MIR pair and lowers it to explicit FMA, so a relaxed consumer cannot absorb a strict multiply;
- no benchmark, client build, vector-width target, or speedup threshold defines correctness.

Verification:

- independent adversarial full-diff review completed; all P1 strategy findings were closed by reopening the owning matrix axes, and the final P2 example mismatch was fixed and recorded;
- `git diff --check` passed;
- `scripts/pre-pr.sh --docs-only` passed on the exact pushed SHA in 0.15 s;
- CI completed in about 10 s: the docs-only classifier skipped the Linux/macOS compiler matrix and PostgreSQL service shards as intended; required aggregate checks passed;
- no Cargo build/test, Clippy, benchmark, full suite, or align-llm build was run for this Markdown-only design PR.

Remaining work:

- implement plan 77 from parser through checked HIR/MIR, explicit FMA selection, LLVM reassociation, formatter, and interface-body rechecking;
- run only the focused syntax, semantic/HIR/MIR, transparent-wrapper, exact-flag/FMA-pair, interface, whole/per-unit, and strict-bit owners defined by the plan;
- perform optional align-llm performance measurements after consumer adoption; they are not provider acceptance gates.

**Comment by sanohiro** [comment]
Completed by #1145 (squash merge `d6a6342106b171a03074bff5a189acca244fb4ac`).

Acceptance criteria closed:
- `float(reassoc)` and `float(contract)` are explicit value-producing lexical scopes; the default and every callable root remain strict.
- Checked HIR retains and authenticates the exact lexical mode, including generic/interface-carried bodies and whole/per-unit compilation.
- Eligible scalar/vector add/sub/mul and all five existing sum/dot forms carry only the named permission.
- Contraction selects an explicit FMA only for a direct pair whose multiply and consumer both permit it, preserves source evaluation order, and emits no raw LLVM `contract`, `nnan`, `ninf`, `nsz`, `arcp`, `afn`, or `fast`.
- Ownership, borrow, region, replay, loop-fact, byte-analysis, malformed-HIR, and ABI paths treat the scope transparently and fail closed.

Verification:
- focused scalar/vector/reduction, checked-HIR mutation, interface round-trip, ownership/borrow, and whole/per-unit owners passed locally; test bodies were 0.00-0.82 s after compilation
- `scripts/test-pr.sh`, workspace Clippy, and the panic/lossy-cast ratchet passed in final-SHA preflight
- required CI passed on Linux ARM64 (6m05s), macOS Apple Silicon (9m57s), Linux x86_64 (11m06s), and all four PostgreSQL shards (9m01s-10m52s)
- no benchmark, nightly full suite, exhaustive numeric corpus, or align-llm build was used as a provider correctness gate

Remaining work:
- align-llm still needs to adopt the lexical scopes at consumer-selected hot paths and rerun its greedy/sampler, object/disassembly, and ordered bit-identity measurements. That consumer-owned work is tracked in Request 110 and does not leave an Align implementation acceptance item open.

GitHub now reports this issue as `CLOSED` with reason `COMPLETED`.
